relay-server/src/extractors/decoder.rs

```diff
@@ -0,0 +1,212 @@
+    /// case, the buffer contains the decoded payload up to the limit. Returns `Err` if there was an
+    /// error decoding.
+    pub fn decode(&mut self, bytes: Bytes) -> io::Result<bool> {
+        match &mut self.inner {
+            DecoderInner::Identity(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Br(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Gzip(inner) => write_overflowing(inner, &bytes),
+            DecoderInner::Deflate(inner) => write_overflowing(inner, &bytes),
+        }
+    }
+
+    /// Returns decoded bytes from the Decoder's buffer.
+    ///
+    /// This can be called at any time during decoding. However, the limit stated during
+    /// initialization remains in effect across the entire decoded payload.
+    pub fn take(&mut self) -> Bytes {
+        match &mut self.inner {
+            DecoderInner::Identity(inner) => inner.take(),
+            DecoderInner::Br(inner) => inner.get_mut().take(),
+            DecoderInner::Gzip(inner) => inner.get_mut().take(),
+            DecoderInner::Deflate(inner) => inner.get_mut().take(),
+        }
+    }
+}
+
+impl fmt::Debug for Decoder {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.inner {
+            DecoderInner::Identity(inner) => f.debug_tuple("Identity").field(inner).finish(),
+            DecoderInner::Br(_inner) => f.debug_tuple("Br").finish(),
+            DecoderInner::Gzip(inner) => f.debug_tuple("Gzip").field(inner).finish(),
+            DecoderInner::Deflate(inner) => f.debug_tuple("Deflate").field(inner).finish(),
+        }
+    }
+}
+
+/// A payload based on [`SharedPayload`] that decompresses the request body on-the-fly.
+///
+/// Note that this stream is not fused.
+#[derive(Debug)]
+pub struct DecodingPayload {
+    payload: SharedPayload,
+    decoder: Decoder,
+}
+
+impl DecodingPayload {
+    /// Creates a decoding payload, resolving chunks of uncompressed request payload.
+    pub fn new<S>(request: &HttpRequest<S>, limit: usize) -> Self {
+        Self {
+            payload: SharedPayload::get(request),
+            decoder: Decoder::new(request, limit),
+        }
+    }
+}
+
+impl Stream for DecodingPayload {
+    type Item = <SharedPayload as Stream>::Item;
+    type Error = <SharedPayload as Stream>::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let encoded = match self.payload.poll() {
+            Ok(Async::Ready(Some(encoded))) => encoded,
+            Ok(Async::Ready(None)) => return Ok(Async::Ready(None)),
+            Ok(Async::NotReady) => return Ok(Async::NotReady),
+            Err(error) => return Err(error),
+        };
+
+        if self.decoder.decode(encoded)? {
+            Err(PayloadError::Overflow)
+        } else {
+            let chunk = self.decoder.take();
+            Ok(Async::Ready(if chunk.is_empty() {
+                None
```

Comment:
Should we `continue` looping here rather than exit? I wonder if we should anticipate a case where a chunk of the payload doesn't yield new decompressed bytes, but the original payload still gives more input.

Reply:
I have no clue what the contract is. I'd rather prematurely end the stream and fail the request parsing than potentially infinite-loop